iconColor is now modifiable via the constructor - InputDecoration enhancement

md-weber opened the pull request:
## Description
The default colour of InputDecoration icon can now be modified with the property iconColor. 

## Related Issues
Closes https://github.com/flutter/flutter/issues/48882

## Tests
I added the following tests:
- InputDecorator should take iconColor into account
- InputDecorator disabledColor should be set

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

md-weber committed:
added the iconColor to input_decorator.dart and tested it

md-weber committed:
removed a clrt

md-weber commented:
I had a suggestion that I could add here also the iconColor on Hover and Disabled state. 
https://github.com/flutter/flutter/issues/48882#issuecomment-659192476

@HansMuller @shihaohong It would be great if you could give me some insights if this is a wished behaviour.

AyushBherwani1998 commented:
Related Issue: https://github.com/flutter/flutter/issues/36248

TimWhiting commented:
@md-weber I was watching your youtube video where you create this PR. 
It looks like you got rid of the disabled color part of _getDefaultIconColor unintentionally.
Take a look at the red (left) side of the diff.
Below is the snippet that you removed unintentionally.
```dart
if (!decoration.enabled && !isFocused)	 
      return themeData.disabledColor;
```

md-weber commented:
Great catch, thank you!

md-weber committed:
reverted the disabledColor and added test

reverted the disabled color back into it and created a test for it

md-weber committed:
Merge remote-tracking branch 'origin/master' into md-weber/issue-48882-input-decoration-icon-color

Hixie commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -2637,6 +2643,9 @@ class InputDecoration {
   /// See [Icon], [ImageIcon].
   final Widget icon;
 
+  /// If specified, the icon color used for the default [icon] color
```
this documentation is highly unhelpful. I can tell it's the icon color, the name is "icon color". :-)

This should say things like what the default is, how it's applied, whether it can be null, when it takes effect and when it gets overridden, etc.

Hixie replied:
see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#documentation-dartdocs-javadocs-etc

md-weber replied:
Thank you for the link, I improved the documentation and hope that it fits now better to the expectations.

Hixie commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -2075,6 +2075,10 @@ class _InputDecoratorState extends State<InputDecorator> with TickerProviderStat
     if (!decoration.enabled && !isFocused)
       return themeData.disabledColor;
 
+    if(decoration.iconColor != null){
```
nit: need spaces around the `(...)`

Hixie replied:
(match the style used elsewhere in the file)

darrenaustin commented on packages/flutter/test/material/input_decorator_test.dart:
```diff
@@ -1449,6 +1451,46 @@ void main() {
+      )
+    );
+
+    final IconTheme iconTheme = tester.firstWidget(
```
A possibly cleaner way of getting the theme would be: 

```dart
    final Element icon = tester.element(find.byIcon(Icons.add_circle));
    final IconTheme iconTheme = icon.findAncestorWidgetOfExactType<IconTheme>();
```

This could apply to the other test below as well.

md-weber replied:
Thanks for that, I changed the behaviour in all tests and added some other tests.

darrenaustin reviewed (changes requested):
My apologies at taking so long to review this. Thanks for the contribution and taking the time to add tests. 

It looks good.  I agree with @hixie's comments below and added a suggestion of my own for the tests. If you can address these small issues, I think we will be good to go.

md-weber commented:
> My apologies at taking so long to review this. Thanks for the contribution and taking the time to add tests.
> 
> It looks good. I agree with @Hixie's comments below and added a suggestion of my own for the tests. If you can address these small issues, I think we will be good to go.

No worries and thanks for your review, I will work on the weekend on that issue. Could you also check the new proposals for the issue, therefore we could enhance the iconColor even further?

https://github.com/flutter/flutter/issues/48882#issuecomment-660012809

md-weber committed:
Merge remote-tracking branch 'origin/master' into md-weber/issue-48882-input-decoration-icon-color

md-weber committed:
wrote test for DecorationIconColors and improved the API to make it possible to change iconColors for TextFields

md-weber committed:
added proper documentation to the DecorationIconColors

md-weber commented:
Thanks for your fantastic feedback. I added now the solution from the Issue that I had in my mind, here we have the possibility to change all the different states of the Icon, this makes it more customizable. If the `DecorationIconColors` class is not existent, it will automatically fall back to the default Theme colours.

Therefore I added several tests that are related to a not existing DecorationIconColors class and for each case, there is a test that if the iconColor for a specific behaviour is set.

md-weber committed:
removed trailing space

md-weber committed:
changed the documentation

md-weber committed:
Merge remote-tracking branch 'remote/master' into md-weber/issue-48882-input-decoration-icon-color

darrenaustin commented on packages/flutter/lib/src/material/input_decorator.dart:
```diff
@@ -468,6 +468,56 @@ class _HelperErrorState extends State<_HelperError> with SingleTickerProviderSta
+/// See also:
+///
+/// * [InputDecoration] where it is used
+class DecorationIconColors {
```
This looks good. However, we have been trying to transition to using `MaterialStateProperty<Color>` for things like this. It allows us to have a single property instead of multiple properties for each state. It also allows the application to customize the state presentation to what ever level they like. If the app wants to specify a color for "hover and disabled" in addition to just hover or disabled, the MaterialStateProperty would allow them to do this.

Can you take a look at this class and see if it would make sense to use it instead of `DecorationIconColors`? To see an example of how it is being used, take a look at the newish `ButtonStyle` class which makes extensive use of this.

md-weber replied:
No worries regarding the delay. Great idea, thank you. I did not know that this class exists 👍. I will have a look, and I can come up with a new solution.

darrenaustin reviewed (changes requested):
@md-weber Thanks again for this contribution. Really sorry for the delay in getting it reviewed. It looks great. My only issue is one that I raise below. It would be nice if we could take advantage of the `MaterialStateProperty<Color>` instead of introducing a new class to represent all the color states. Can you take a look and see if this makes sense to you? Thx!

darrenaustin commented:
@md-weber, I hope you don't mind, but I am going to close this to get it off the review pile. Please reopen or open a new PR when you have the chance to rework this. Thanks again for your work on this.
